PR demonstrating WebAssembly support

NWilson opened the pull request:
I know that Musl doesn't accept changes via pull-requests!

This is just a way of showing the current status of my work.

Go to the diff to see what changes I made to support Wasm.

fabled committed:
implement strftime padding specifier extensions

notes added by maintainer:

the '-' specifier allows default padding to be suppressed, and '_'
allows padding with spaces instead of the default (zeros).

these extensions seem to be included in several other implementations
including FreeBSD and derivatives, and Solaris. while portable
software should not depend on them, time format strings are often
exposed to the user for configurable time display. reportedly some
python programs also use and depend on them.

NWilson committed:
[Wasm] first cut of Wasm support - most basic libc features working!

jow- committed:
fix endian errors in arpa/nameser.h due to failure to include endian.h

kaniini committed:
implement the fopencookie extension to stdio

notes added by maintainer:

this function is a GNU extension. it was chosen over the similar BSD
function funopen because the latter depends on fpos_t being an
arithmetic type as part of its public API, conflicting with our
definition of fpos_t and with the intent that it be an opaque type. it
was accepted for inclusion because, despite not being widely used, it
is usually very difficult to extricate software using it from the
dependency on it.

calling pattern for the read and write callbacks is not likely to
match glibc or other implementations, but should work with any
reasonable callbacks. in particular the read function is never called
without at least one byte being needed to satisfy its caller, so that
spurious blocking is not introduced.

contracts for what callbacks called from inside libc/stdio can do are
always complicated, and at some point still need to be specified
explicitly. at the very least, the callbacks must return or block
indefinitely (they cannot perform nonlocal exits) and they should not
make calls to stdio using their own FILE as an argument.

richfelker committed:
adjust fopencookie structure tag for ABI-compat

stdio types use the struct tag names from glibc libio to match C++
ABI.

NWilson committed:
[Wasm] Add math.h overrides using Wasm intrinsics

NWilson committed:
[Wasm] Add __syscall_mmap implementation

NWilson committed:
[Wasm] Use LDFLAGS when testing linker

NWilson committed:
remove unused explicit dependency rules for crti/crtn

notes by maintainer:

commit 2f853dd6b9a95d5b13ee8f9df762125e0588df5d added these rules
because the new system for handling arch-provided replacement files
introduced for out-of-tree builds did not apply to the crt tree.

commit 63bcda4d8f4074e9d92ae156afd0dced6e64eb65 later adapted the
makefile logic so that the crt and ldso trees go through the same
replacement logic as everything else, but failed to remove the
explicit rules that assumed the arch would always provide asm
replacements.

in addition to cleaning things up, removing these spurious rules
allows crti/crtn asm to be omitted by an arch (thereby using the empty
C files instead) if they are not needed.

NWilson committed:
[Wasm] Add wasm.syms file

ncopa committed:
use the name UTC instead of GMT for UTC timezone

notes by maintainer:

both C and POSIX use the term UTC to specify related functionality,
despite POSIX defining it as something more like UT1 or historical
(pre-UTC) GMT without leap seconds. neither specifies the associated
string for %Z. old choice of "GMT" violated principle of least
surprise for users and some applications/tests. use "UTC" instead.

NWilson committed:
[Wasm] Remove temporary debugging #define

ncopa committed:
fix sysconf for infinite rlimits

sysconf should return -1 for infinity, not LONG_MAX.

sunfishcode commented on arch/wasm/bits/limits.h:
```diff
@@ -0,0 +1,17 @@
+ * allocations.  At 32KiB page size, we have <20% RAM
+ * waste which seems a bit more reasonable.  All a bit arbitrary.
+ */
+#define PAGE_SIZE 32768
```
(Random drive-by review comment...)

This property of page sizes may not always hold; it is possible that wasm could add [mmap-like features in the future](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md#finer-grained-control-over-memory), in which case PAGE_SIZE should reflect the granularity of those operations.

I realize that you don't want to mix in bigger changes in this PR, but if malloc's heuristics are suboptimal for 64KiB page sizes, it would affect other architectures as well, including some arm64 platforms, so ideally the solution should be to fix malloc's heuristics.

sunfishcode commented on arch/wasm/bits/alltypes.h.in:
```diff
@@ -0,0 +1,33 @@
+TYPEDEF float float_t;
+TYPEDEF double double_t;
+
+TYPEDEF long time_t;
```
`long` is 32-bit on wasm32, so using it for `time_t` is susceptible to the [2038 bug](https://en.wikipedia.org/wiki/Year_2038_problem). Since wasm has a full set of 64-bit operations, I suggest making `time_t` 64-bit.

NWilson replied:
Yes, what was I thinking!?! The ABI is supposed to match the Linux x32 ABI, this must have snuck in from the x86 one by mistake.

I'll do a diff between arch/x32 and arch/wasm and make sure everything matches where possible.

(By comparing the syscall numbers from all the different architectures, I've deduced that Emscripten is currently trying to emulate the x86 ABI in its syscall interface, or at least seems to have copied the definitions from x86 most extensively. I think moving to the x32 definitions is a good move at this point.)

NWilson commented on arch/wasm/bits/syscall.h.in:
```diff
@@ -0,0 +1,325 @@
+// The list of syscalls is taken from x32, which contains a sensible list of
+// "modern" syscalls.
+
+#define SYSCALL_ACTION(x) long __syscall_##x(long arg1, ...);
```
Some syscalls take a variable number of arguments (like SYS_open, and SYS_mremap).

I thought it would be cunning to therefore give syscalls a varargs signature. (NB. The way static linking for syscalls has to work is that all syscalls must have the same signature. See how support for `syscall()` and `pthread_cancel()` was implemented by cunningly using each syscall's function pointer as its syscall number! Hat-tip to John Starks at Microsoft.)

Making them varargs is a silly idea, I should just make all take six normal arguments, and pass zeros for the ones that don't matter.

sunfishcode committed:
Make FORCE_EVAL a no-op on wasm.

WebAssembly doesn't have user-accessible floating point exceptions, so
there's no advantage to forcing floating-point expression evaluations.

NWilson committed:
[Wasm] Synchronise a few more definitions with x32 rather than x86

NWilson committed:
[Wasm] Make syscall interface non-varargs

sunfishcode commented on arch/wasm/bits/stdint.h:
```diff
@@ -0,0 +1,31 @@
+typedef int32_t int_fast16_t;
```
This definition of int_fast16_t differs from clang's, but I think it's better. I filed [this bug](https://bugs.llvm.org/show_bug.cgi?id=35582) to track the difference.

sunfishcode commented on src/internal/wasm/wasm_init_data.c:
```diff
@@ -0,0 +1,26 @@
+
+static const char* const progname = "/a.out";
+
+struct wasm_init_data_t wasm_init_data = {
```
Should this be named `__wasm_init_data` to avoid poluting the user namespace? Also, I think this can be `const` (though wasm doesn't have read-only memory yet).

NWilson replied:
Oops, hiding with underscores done.

Making `const` unfortunately not possible; Musl functions like `_start_c` from crt1.c expect a non-const pointer. Making const only to cast it away seems pointless.

I think for time being this sort of ELF emulation is OK. It's not too burdensome, and it's emulating a well-defined API for interaction with Musl.

If `_start_wasm` does become a standard thing, we should maybe add argv/envp to it as arguments.

Finally, a getrandom opcode in Wasm (`__builtin_wasm_getrandom`) would be nice :)

sunfishcode commented on src/math/wasm/truncf.c:
```diff
@@ -0,0 +1,6 @@
+__attribute__((const)) float truncf(float x)
+{
+	return __builtin_truncf(x);
+}
```
You can also optimize `nearbyint`/`nearbyintf` and `rint`/`rintf` using `__builtin_nearbyint`/`__builtin_nearbyintf` and `__builtin_rint`/`__builtin_rintf`, respectively, which codegen to `f64.nearest`/`f32.nearest`. (This is possible for rint because wasm doesn't have user-accessible floating-point exceptions, and for nearbyint and rint because wasm doesn't have user-modifiable rounding modes.

NWilson replied:
Thanks, added those too.

I did a bit of trial-and-error to work out which math builtins could be optimised this way. Adding `__attribute__((const))` was a bit of trial-and-error. Musl doesn't set errno for any maths functions, so it should be safe/consistent.

I was hoping to turn fmin/fmax into `f64.min`, but couldn't seem to get it to be emitted by `__builtin_fmax`. Those builtins are tricky, because if they don't lower to the expected native opcode, then they generate a recursive call to the libc function!

In any case, these routines in libc are pretty unimportant. LLVM replaces calls to these libc functions with the native opcodes anyway, so they won't/shouldn't actually be called here unless you're calling them by pointer.

NWilson replied:
Can you think of a reason why this body doesn't emit a nice `f64.min` instruction:

`__attribute__((const)) double fmin(double x, double y) { return __builtin_fmin(x,y); }`

I'm not confident in saying it's a bug in the WebAssembly frontend, or whether my expectation is wrong, but it is surprising that so many of the other maths builtins do work.

sunfishcode replied:
When one operand is NaN, wasm's `min` and `max` return NaN, while C's `fmin` and `fmax` return the other operand. Both forms are useful, so I expect that in the future wasm will have opcodes for both forms. As a historical note, the reason why wasm initially left out the latter form because IEEE 754-2008's definition of this form had surprising behavior on signaling NaN (see [here](https://2pi.dk/2016/05/ieee-min-max) for a description). The IEEE 754 committee has recently recognized this as an error, and drafts of the upcoming 754-2018 have removed that definition and provided a new corrected one.

NWilson replied:
I see, thanks!

sunfishcode commented on src/internal/wasm/wasm_init_data.c:
```diff
@@ -0,0 +1,26 @@
+#endif
+		AT_NULL,     0
+	}
+};
```
Random musing: I realize that this serves a purpose right now, though at some point we may want to explore other options. Portable code won't likely use interfaces like `getauxv` in the first place, for wasm-specific code there are better ways to obtain most of this information, and for musl initialization, there isn't all that much code that needs these, and some of it needs auxv fields that aren't supported (AT_PHDR etc.).

NWilson committed:
[Wasm] hide wasm_init_data symbol with underscores

NWilson committed:
[Wasm] Add another couple of optimised math functions

sunfishcode commented on arch/wasm/bits/posix.h:
```diff
@@ -0,0 +1,2 @@
+#define _POSIX_V6_LP64_OFF64  1
+#define _POSIX_V7_LP64_OFF64  1
```
wasm32 is ILP32, so it looks like these should be ILP32_OFFBIG rather than LP64_OFF64, for wasm32.

NWilson replied:
Hm, I just copied that blindly from the x32 port - which is also ILP32, surely? So it looks like a bug in `arch/x32/bits/posix.h` that I didn't notice or think about. `off_t` is complicated, I was just hoping I wouldn't have to check out the specs for it!

I'll raise it tomorrow on the Musl mailing list for the x32 port.

NWilson committed:
fix x32 unistd macros to report as ILP32 not LP64

NWilson committed:
[Wasm] Use ILP32 rather than LP64 macros on Wasm

NWilson commented:
On an aside, I'm currently having a go at a "JavaScript linker".

Emscripten currently does this using a very awkward `<symbol>__deps` annotation, which feels a bit odd to me. I'm trying to build something simple using nodejs, that does the following steps:

1. Load WebAssembly file and list of Javascript files specified on the commandline
2. For the Wasm module, grab its dependencies from its import list
3. For the JS files, parse them using Shift AST, to get a list of symbol definitions, and extract all referenced variables using Shift's scope analyser, ie automatically build the deps list
4. Output a Javascript module that pulls in the JS symbols that are used, and loads the Wasm module with the symbols it imports. JS symbols are all in a shared scope, so can reference each other regardless of whether they were pulled in by the Wasm module or other imports.

I don't think there's anything quite like that at the moment? Emscripten's Python module loader has quite a lot of asm.js going on, and it would be nice to have a short & clean linker/JS-module-builder that just knows how to assemble the Javascript side of a Wasm module.

Then there'd need to be a libc.js (derived from Emscripten's existing library.js) that implements some sane subset of the Musl syscalls.

richfelker committed:
add ibm1047 codepage (ebcdic representation of latin1) to iconv

richfelker committed:
fix data race in at_quick_exit

aside from theoretical arbitrary results due to UB, this could
practically cause unbounded overflow of static array if hit, but
hitting it depends on having more than 32 calls to at_quick_exit and
having them sufficiently often.

NWilson committed:
[Wasm] Fix __builtin_unreachable misuse

NWilson committed:
[Wasm] Hardcode float constants to avoid ABI depending on the compiler

NWilson committed:
[Wasm] Update some comments

NWilson committed:
[Wasm] Add functioning dummy for pthread_self()

NWilson committed:
[Wasm] Remove bits/errno.h override, better to fix ELAST problem in libcxx

NWilson commented on arch/wasm/bits/float.h:
```diff
@@ -0,0 +1,20 @@
+#define LDBL_MAX 1.18973149535723176508575932662800702e+4932L
+#define LDBL_EPSILON 1.92592994438723585305597794258492732e-34L
+
+#define LDBL_MANT_DIG 113
```
@sunfishcode, do you know how Wasm has 80-bit long-double support? I would have expected long-double to be the same as double, and was surprised to find that it's not. Does clang emulate the higher-precision, or is it fictitious?

I got these constants here by dumping Clang's predefined macros (as requested by Rich Felker in his review on the Musl mailing list).

sunfishcode replied:
Clang is currently configured to make long-double be a 128-bit IEEE-754 quad-precision type. It is supported through software emulation in compiler-rt (it calls __addtf3 for addition, and so on).

The decision to do this is debatable. On one hand, making long double the same as double makes long double effectively useless, and making it longer makes it at least useful for some purposes (even if they are fairly narrow). On the other, musl's printf promotes all floating-point values to `long double` to avoid duplicating code, and this means that all floating-point formating goes through slow software-emulated paths, and it increases code size for any program that calls printf.

I expect we'll eventually have custom versions of printf that will disable disable various bits of floating-point support, because many applications won't need everything (and the nature of printf format strings makes it difficult to dce unused formatting code), and if we do, this would significantly mitigate the problems. However, the utility of 128-bit long double is admittedly obscure, so it's not completely certain that the benefits outweigh the costs.

NWilson replied:
When you say "Clang is currently configured", I assume that's specific to the WebAssembly backend? Given that GCC on x86 doesn't give you 128-bit quad-precision floats, the need for that level of precision seems remote.

I don't know - maybe people really would welcome having that high precision available in Wasm via Clang. But, to me it just seems like a pitfall.

Something to think about, thanks for explaining. I'd vote to have double match long-double on Wasm, at least until the demand for it is strong enough. Since we don't yet have dynamic linkage, changing the ABI isn't too big a deal.

NWilson replied:
I've found https://github.com/kripken/emscripten/issues/4340 now. I see Derek was thinking of going back to 64-bit long-double, but that doesn't seem to have happened. Oh well - up to you guys!

NWilson committed:
[Wasm] Exclude internal syscall from exported syms list

richfelker committed:
fix endian errors in netinet/icmp6.h due to failure to include endian.h

NWilson committed:
[Wasm] Move Wasm initialisation to use Dan's global-ctor entrypoint

NWilson committed:
[Wasm] Switch to using Dan's global-constructor approach to initialisation

NWilson committed:
[Wasm] rename __heap_bottom to __heap_base

NWilson committed:
Fix use of weakly-defined address, not variable

NWilson committed:
Add support for thread-local data to Wasm

NWilson committed:
Wasm: remove some x86 syscalls

NWilson committed:
Wasm: add __syscall implementation for the sake of __syscall_cp

NWilson committed:
Wasm: Fix build by stubbing out x86-specific calls

skierpage commented:
> On an aside, I'm currently having a go at a "JavaScript linker".

FYI, as is https://github.com/lrz/mono-wasm-libc  (I apologize if this is spam; since musl master is so disconnected, it's difficult to track the state of all its forks implementing a wasm target. Thanks for all you do!)
